### DISCUSSION TITLE
How to close the Kyuubi session and spark cluster?

### DISCUSSION THREAD

**Discussion by HaoYang670** [apache/kyuubi#4449] [Q&A] [answered]
I am currently using [JayDeBeApi](https://pypi.org/project/JayDeBeApi/#usage) to connect to kyuubi service on K8S to run some SQL queries on Spark. The usage is like this
```python
import jaydebeapi

connection = jaydebeapi.connect("org.apache.kyuubi.jdbc.KyuubiHiveDriver", <kyuubi service endpoint>, jars = [<kyuubi_jdbc_hive_jar>])
cursor = connection.cursor()
cursor.execute(<sql query>)
cursor.close()
connection.close()
```

I meet a problem when doing this, which I cannot close the Kyuubi session and the Spark cluster. 
Is there a way to close them using the JDBC API?

**Comment by yaooqinn** [comment, marked as answer]
If you want to terminate the spark app when closing the connection, you need to set the share level to CONNECTION, see https://kyuubi.readthedocs.io/en/master/deployment/engine_lifecycle.html.

**Reply by HaoYang670** [reply]
Thank you @yaooqinn. But I will use Kyuubi in `USER` level and `GROUP` level unfortunately. Is there any way that I can close the Spark cluster in these 2 modes? (Can RESTful API achieve this?)

**Reply by yaooqinn** [reply]
If you choose USER / GROUP share level, engines will share across connections, it's unsafe to shut down in one connection. However, you can use kyuubi.operation.language=scala to call spark.stop

**Reply by HaoYang670** [reply]
Thank you. I'd like to try this. 
But how can I use this (especially from Python side)? I don't find a doc except this: https://kyuubi.readthedocs.io/en/v1.6.1-incubating/deployment/settings.html#operation

Is there any example or tutorial of how to use it?

**Reply by pan3793** [reply]
Just execute two statements(same as normal SQL) sequentially

**Reply by HaoYang670** [reply]
Thank you @pan3793.
```python
cursor.execute("SET kyuubi.operation.language=scala")
cursor.execute("spark.stop")
```
This can delete the Spark Context, but raise an error
```
org.apache.kyuubi.jdbc.hive.org.apache.kyuubi.jdbc.hive.KyuubiSQLException: org.apache.kyuubi.jdbc.hive.KyuubiSQLException: org.apache.kyuubi.KyuubiSQLException: Error operating ExecuteStatement: org.apache.thrift.transport.TTransportException: java.net.SocketException: Broken pipe (Write failed)
```
How can I fix it?

Besides, I also want to delete the Kyuubi driver. Is there a similar way to do it?

**Reply by pan3793** [reply]
it's kind of hacky way, so exception is expected

**Comment by wang-shengcai** [comment]
Another thread shuts down the currently running connection, which should do just that

**Comment by HaoYang670** [comment]
Setting `kyuubi.operation.language=scala` then calling `spark.stop` works for me.
Thank you all for the help!
